### PR TITLE
#1681 Added the mentioned algorithms under string algorithms

### LINKED ISSUE #1681: [Feature]: Addition of more algorithms under String Algorithms
### Feature Name

Addition of more algorithms under String Algorithms

### Feature Description

I will add the following algorithms under string algorithms:
1)SBOM (Set Backward Oracle Matching)
2)Galil-Seiferas Algorithm
3)DAWG-Match Algorithm

### Motivation

Newer and diversity of algorithms helps the user to understand.

### Implementation Suggestions (Optional)

_No response_

### Feature Type

New Algorithm

### Does this feature require additional resources?

- [X] Additional Documentation
- [ ] Third-Party Libraries
- [ ] UI/UX Design
- [ ] Other

### References (Optional)

_No response_

### PR DESCRIPTION
**Pull request by pratheekv39**
### Description
Added the mentioned algorithms under string algorithms
1)SBOM (Set Backward Oracle Matching)
2)Galil-Seiferas Algorithm
3)DAWG-Match Algorithm

Fixes #1681

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

@ajay-dhangar Kindly review and merge the PR .... Thanks :)

### CODE REVIEW

File: docs/algorithms/string-algorithms/dawg-match-algorithm.md [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+           state = dawg[state].transitions[text[i]];
+           
+           if (dawg[state].isTerminal) {
+               matches.push_back(i - pattern.length() + 1);
+           }
+       }
+       
+       return matches;
+   }
+};
```
Comment (ajay-dhangar) [inline docs/algorithms/string-algorithms/dawg-match-algorithm.md:210]: close your code with ```` ``` ````

File: docs/algorithms/string-algorithms/galil-seiferas-algorithm.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+            } else {
+                // Optimized shifting based on local period
+                int period = computeLocalPeriod(pattern, 0, j - 1);
+                i += max(1, j - period);
+            }
+        }
+        
+        return matches;
+    }
+};
```
Comment (ajay-dhangar) [inline docs/algorithms/string-algorithms/galil-seiferas-algorithm.md:155]: close your code example with ```` ``` ````

File: docs/algorithms/string-algorithms/sbom-algorithm.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+            if (j == pos - pattern.length()) {
+                matches.push_back(pos - pattern.length() + 1);
+            }
+            
+            pos += pattern.length();
+        }
+        
+        return matches;
+    }
+};
```
Comment (ajay-dhangar) [inline docs/algorithms/string-algorithms/sbom-algorithm.md:181]: close your code example with ```` ``` ````

**Review by ajay-dhangar — CHANGES_REQUESTED** [review]
Update your code based on suggestions

File: docs/algorithms/string-algorithms/sbom-algorithm.md (hunk 3 above)
Comment (pratheekv39) [reply docs/algorithms/string-algorithms/sbom-algorithm.md:181]: @ajay-dhangar There is a problem with repo . It is causing sync and update problems . Kindly, fix this

**Comment by pratheekv39** [comment]
@ajay-dhangar Can you please fix the error with the repo . Can't sync and update ..... Throwing the same error 

 * branch              stringalgorithms -> FETCH_HEAD
error: invalid path 'docs/Fenwick Tree /FenwickTree.md'
error: invalid path 'docs/algorithms/Searching Algorithms/A*-Search.md'
error: invalid path 'docs/algorithms/Searching Algorithms/AO*-Search.md'
Merge with strategy ort failed.

**Comment by ajay-dhangar** [comment]
> @ajay-dhangar Can you please fix the error with the repo . Can't sync and update ..... Throwing the same error
> 
> * branch              stringalgorithms -> FETCH_HEAD
>   error: invalid path 'docs/Fenwick Tree /FenwickTree.md'
>   error: invalid path 'docs/algorithms/Searching Algorithms/A*-Search.md'
>   error: invalid path 'docs/algorithms/Searching Algorithms/AO*-Search.md'
>   Merge with strategy ort failed.

Reclone our repo with latest commit

**Comment by pratheekv39** [comment]
@ajay-dhangar Nope, it still shows the same issue , now a new file's naming convention is also causing issues. Please remove spaces from them , or special characters from these files to follow the proper naming convention that GitHub can recognize properly.

error: invalid path 'docs/Fenwick Tree /FenwickTree.md'
error: invalid path 'docs/algorithms/Searching Algorithms/A*-Search.md'
error: invalid path 'docs/algorithms/Searching Algorithms/AO*-Search.md'

(Extra Space in Fenwick Tree Directory, and the remaining two file's naming convention - Remove * from them as it is not recognized )

**Comment by pratheekv39** [comment]
@ajay-dhangar Just a heads-up, filenames like AO*-Search.md are technically invalid on Windows, which could cause issues for anyone not on macOS. However, special characters like *  would bypass naming conventions on Linux and macOS based systems .....It’s always good to avoid special characters(and also blank spaces) if we want the repo to work for everyone.

**Comment by ajay-dhangar** [comment]
> @ajay-dhangar Nope, it still shows the same issue , now a new file's naming convention is also causing issues. Please remove spaces from them , or special characters from these files to follow the proper naming convention that GitHub can recognize properly.
> 
> error: invalid path 'docs/Fenwick Tree /FenwickTree.md' error: invalid path 'docs/algorithms/Searching Algorithms/A*-Search.md' error: invalid path 'docs/algorithms/Searching Algorithms/AO*-Search.md'
> 
> (Extra Space in Fenwick Tree Directory, and the remaining two file's naming convention - Remove * from them as it is not recognized )

Sure I'll fix this issue So, wait for some time ⌚

**Pull request closed without merging**